src/checker/checker.py

```diff
@@ -0,0 +1,370 @@
+from __future__ import annotations
+
+from abc import abstractmethod
+from dataclasses import dataclass
+from enum import IntEnum
+from typing import TYPE_CHECKING, cast
+
+if TYPE_CHECKING:
+    from collections.abc import Iterator
+
+
+class Instruction(IntEnum):
+    List = 1
+
+    # Patterns
+    EVar = 2
+    SVar = 3
+    Symbol = 4
+    Implication = 5
+    Application = 6
+    Mu = 7
+    Exists = 8
+
+    # Meta Patterns
+    MetaVar = 9
+    ESubst = 10
+    SSubst = 11
+
+    # Axiom Schemas
+    Prop1 = 12
+    Prop2 = 29  # TODO: Fix numbering
+    Prop3 = 30  # TODO: Fix numbering
+    Quantifier = 13
+    PropagationOr = 14
+    PropagationExists = 15
+    PreFixpoint = 16
+    Existance = 17
+    Singleton = 18
+
+    # Inference rules
+    ModusPonens = 19
+    Generalization = 20
+    Frame = 21
+    Substitution = 22
+    KnasterTarski = 23
+
+    # Meta Incference rules
+    InstantiateSchema = 24
+
+    # Stack Manipulation
+    Pop = 25
+
+    # Memory Manipulation
+    Save = 26
+    Load = 27
+
+    # Journal Manipulation
+    Publish = 28
+
+
+class Pattern:
+    @abstractmethod
+    def instantiate(self, var: MetaVar, plug: Pattern) -> Pattern:
+        ...
+
+
+@dataclass(frozen=True)
+class EVar(Pattern):
+    name: int
+
+    def instantiate(self, var: MetaVar, plug: Pattern) -> EVar:
+        return self
+
+
+@dataclass(frozen=True)
+class SVar(Pattern):
+    name: int
+
+    def instantiate(self, var: MetaVar, plug: Pattern) -> SVar:
+        return self
+
+
+@dataclass(frozen=True)
+class Symbol(Pattern):
+    name: int
+
+    def instantiate(self, var: MetaVar, plug: Pattern) -> Symbol:
+        return self
+
+
+@dataclass(frozen=True)
+class Implication(Pattern):
+    left: Pattern
+    right: Pattern
+
+    def instantiate(self, var: MetaVar, plug: Pattern) -> Implication:
+        return Implication(self.left.instantiate(var, plug), self.right.instantiate(var, plug))
+
+    def __repr__(self) -> str:
+        return '(' + str(self.left) + ' -> ' + str(self.right) + ')'
+
+
+@dataclass(frozen=True)
+class Application(Pattern):
+    left: Pattern
+    right: Pattern
+
+    def instantiate(self, var: MetaVar, plug: Pattern) -> Application:
+        return Application(self.left.instantiate(var, plug), self.right.instantiate(var, plug))
+
+
+@dataclass(frozen=True)
+class Exists(Pattern):
+    var: EVar
+    subpattern: Pattern
+
+    def instantiate(self, var: MetaVar, plug: Pattern) -> Exists:
+        return Exists(self.var.instantiate(var, plug), self.subpattern.instantiate(var, plug))
+
+
+@dataclass(frozen=True)
+class Mu(Pattern):
+    var: SVar
+    subpattern: Pattern
+
+    def instantiate(self, var: MetaVar, plug: Pattern) -> Mu:
+        return Mu(self.var.instantiate(var, plug), self.subpattern.instantiate(var, plug))
+
+
+@dataclass(frozen=True)
+class MetaVar(Pattern):
+    name: int
+    fresh: tuple[EVar | SVar, ...]
+    positive: tuple[SVar, ...]
+    negative: tuple[SVar, ...]
```

Comment:
Why do we have both 'positive' and 'negative', when an occurrence of a variable can only be either positive or negative? Is this for performance reasons?

Reply:
We need both to be able to prove the other, since they are only defined mutually. Note that a variable meets both positivity and negativity if it does not occur at all.

Reply:
So if it's in both, it does not occur, if it is in one of them, it is positive/negative, and if it is in neither of them, then it is undefined?